docs/src/main/asciidoc/datastore.adoc

```diff
@@ -985,6 +985,36 @@ If a method's return type is `org.springframework.data.domain.Page`, the returne
 NOTE: Methods that return `Page` execute an additional query to compute total number of pages.
 Methods that return `Slice`, on the other hand, don't execute any additional queries and therefore are much more efficient.
 
+==== Null Handling of Repository Methods
+Java `java.util.Optional` can be used to indicate the potential absence of a return value.
+
+Alternatively, query methods can return the result without a wrapper.
+In that case the absence of a query result is indicated by returning `null`.
+Repository methods returning collections are guaranteed never to return `null` but rather the corresponding empty representation.
+
+==== Nullability Annotations
+You can express nullability constraints for repository methods by using https://docs.spring.io/spring/docs/5.2.5.RELEASE/spring-framework-reference/core.html#null-safety[Spring Framework’s nullability annotations].
+They provide an opt-in null checks during runtime, as follows:
+
+ . `@NonNullApi`: Used on the package level to declare that the default behavior for parameters and return values is to not accept or produce null values.
+
+ . `@NonNull`: Used on a parameter or return value that must not be null.
+
+ . `@Nullable`: Used on a parameter or return value that can be `null` (not needed on a parameter and return value where @NonNullApi applies).
+
+To enable runtime checking of nullability constraints for query methods, you need to activate non-nullability on the package level by using Spring’s `@NonNullApi` in `package-info.java`, as shown in the following example:
```

Comment:
Is there anything you can use at a class level or method level?
Wouldn't `@NonNull` on  a method work?
This sentence makes it sound like you can only make it work on methods by using the package-level annotation which is extremely inflexible and cumbersome.

Reply:
I think `@NonNull` should work on a method without `@NonNullApi`, but for some reason it doesn't. I'll look into that.

Reply:
Let's resolve this before merging.

Reply:
So turns out the class has to be annotated with `javax.annotation.Nonnull` in order for `@NonNull` to work on methods and parameters. This is not documented. I'm not sure if it's bag or undocumented feature.

Reply:
Can you file it as bug/question on the Spring side just to make sure we're not missing anything.

Reply:
I created an issue for Spring Data: https://jira.spring.io/browse/DATACMNS-1705